semgrep/semgrep/semgrep_types.py

```diff
@@ -76,6 +129,10 @@ class BooleanRuleExpression:
     children: Optional[List[Any]] = None
     operand: Optional[str] = None
 
+    # For tests, eg. don't force people to make spans
+    span: Optional[Span] = None
```

Comment:
Why does the BooleanRuleExpression need Span objects?

Reply:
I think it would be less cognitive load to pull out the validation code out of the object into its own function/class and have that thing handle provenance.

Reply:
because at the time of validation, we've lost the original source. In general, keeping spans attached to things is good because it means you can always show the relevant code for errors. (eg. if an individual booleanexpression causes a semgrep-core error)

I could be convinced we should be keeping an external mapping from `BooleanExpression` => `Span` though if you prefer

Reply:
Might be enough to just rename to provenance or something to avoid confusion wrt line ranges of findings and line ranges of the rule provenance. I think if it's not much more work to pull out the validation stuff might be less messy.